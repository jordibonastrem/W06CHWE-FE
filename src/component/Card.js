import "./Card.scss";
import { object, string, number, bool, shape } from "prop-types";

function Card({
  robot: {
    name,
    imageSource,
    features: { speed, endurance, creationDate, isItCute },
  },
}) {
  return (
    <li className={"list-element col-4"}>
      <div className={"card"}>
        <img
          className={"image card-img-top"}
          src={imageSource}
          alt="Robot"
          width={"100%"}
          height={"200px"}
        />

        <div className={"card-body"}>
          <h2 className={"card-title"}>{name}</h2>
          <h3
            className={"card-subtitle mb-2 text-muted"}
          >{`created on ${creationDate}`}</h3>
          <p
            className={".card-text"}
          >{`This robot has a speed of ${speed}, endurance of ${endurance} and it's ${
            isItCute ? "very cute" : "kind of ugly"
          }.`}</p>
          <div className={"d-flex justify-content-around"}>
            <button className={"btn btn-primary"}>Edit</button>
            <button className={"btn btn-danger"}>Delete</button>
          </div>
        </div>
      </div>
    </li>
  );
}

Card.propTypes = {
  robot: shape({
    name: string.isRequired,
    imageSource: string.isRequired,
    features: shape({
      speed: number.isRequired,
      endurance: number.isRequired,
      creationDate: object.isRequired,
      isItCute: bool,
    }).isRequired,
  }).isRequired,
};

export default Card;
