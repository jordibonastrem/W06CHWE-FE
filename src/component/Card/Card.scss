h2 {
  font-size: 1.25rem;
}
h3 {
  font-size: 1rem;
}
.image {
  object-fit: cover;
}

button {
  width: 80px;
}
.login-button {
  width: auto;
}
