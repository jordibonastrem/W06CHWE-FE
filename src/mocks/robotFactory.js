import { Factory } from "fishery";
import { name, image, datatype, lorem } from "faker";

const factoryRobots = Factory.define(({ sequence }) => ({
  id: sequence,
  name: name.firstName(),
  imageSource: image.imageUrl(),
  features: {
    speed: datatype.number({ min: 0, max: 10 }),
    endurance: datatype.number({ min: 0, max: 10 }),
    creationDate: lorem.word,
    isItCute: datatype.boolean(),
  },
}));

export const getRandomRobot = () => factoryRobots.build();
export const getRandomRobots = (total = 5) => factoryRobots.buildList(total);
