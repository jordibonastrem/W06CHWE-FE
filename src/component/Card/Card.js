import "./Card.scss";
import { string, number, bool, shape } from "prop-types";

function Card({ robot }) {
  return (
    <li className={"list-element col-4"}>
      <div className={"card"}>
        <img
          className={"image card-img-top"}
          src={robot.imageSource}
          alt="Robot"
          width={"100%"}
          height={"200px"}
        />

        <div className={"card-body"}>
          <h2 className={"card-title"}>{robot.name}</h2>
          <h3
            className={"card-subtitle mb-2 text-muted"}
          >{`created on ${robot.features.creationDate}`}</h3>
          <p className={".card-text"}>{`This robot has a speed of ${
            robot.features.speed
          }, endurance of ${robot.features.endurance} and it's ${
            robot.features.isItCute ? "very cute" : "kind of ugly"
          }.`}</p>
          <div className={"d-flex justify-content-around"}>
            <button className={"btn btn-primary"}>Edit</button>
            <button className={"btn btn-danger"}>Delete</button>
          </div>
        </div>
      </div>
    </li>
  );
}

Card.propTypes = {
  robot: shape({
    name: string.isRequired,
    imageSource: string.isRequired,
    features: shape({
      speed: number.isRequired,
      endurance: number.isRequired,
      creationDate: string.isRequired,
      isItCute: bool,
    }).isRequired,
  }).isRequired,
};

export default Card;
