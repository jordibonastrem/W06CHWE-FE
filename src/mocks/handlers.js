import { robotPaths } from "../paths/paths";
import { rest } from "msw";
import { getRandomRobots } from "./robotFactory";

export const handlers = [
  rest.get(
    process.env.REACT_APP_URL + robotPaths.get,
    async (req, res, ctx) => {
      const response = res(ctx.json(getRandomRobots()));
      return response;
    }
  ),
];
