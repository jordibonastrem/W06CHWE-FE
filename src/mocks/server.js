import { handlers } from "./handlers";
import { setUpServer } from "./handlers";

export const server = setUpServer(...handlers);
