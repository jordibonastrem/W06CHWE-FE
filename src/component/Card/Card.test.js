import { getRandomRobot } from "../../mocks/robotFactory";
import Card from "./Card";
import { screen, render } from "@testing-library/react";
import ReactTestRenderer from "react-test-renderer";

describe("Given a Card component,", () => {
  let robot;
  beforeEach(() => {
    robot = getRandomRobot();
  });

  describe("When it receives an empty object", () => {
    test("Then it renders!", () => {
      robot = {
        id: "",
        name: "",
        imageSource: "",
        features: {
          speed: NaN,
          endurance: NaN,
          creationDate: "",
          isItCute: false,
        },
      };
      render(<Card robot={robot} />);
    });
  });

  describe("When it recieves a robot object", () => {
    test("Then it renders!", () => {
      render(<Card robot={robot} />);
    });

    test("Then it renders correctly its data", () => {
      render(<Card robot={robot} />);

      const name = screen.getByRole("heading", { name: robot.name });
      const imageSource = screen.getByAltText("Robot");
      const date = screen.getByRole("heading", {
        name: `created on ${robot.features.creationDate}`,
      });
      const paragraph = screen.getByText(
        `This robot has a speed of ${robot.features.speed}, endurance of ${
          robot.features.endurance
        } and it's ${robot.features.isItCute ? "very cute" : "kind of ugly"}.`
      );

      expect(name).toBeInTheDocument();
      expect(imageSource.src).toBe(robot.imageSource);
      expect(paragraph).toBeInTheDocument();
      expect(date).toBeInTheDocument();
    });

    test("Snapshot of the test", () => {
      robot = {
        id: "123",
        name: "Juanito",
        imageSource: "http://localhost:666/",
        features: {
          speed: 3,
          endurance: 5,
          creationDate: "1994",
          isItCute: false,
        },
      };
      const renderedComponent = ReactTestRenderer.create(
        <Card robot={robot} />
      );

      expect(renderedComponent.toJSON()).toMatchSnapshot();
    });
  });
});
