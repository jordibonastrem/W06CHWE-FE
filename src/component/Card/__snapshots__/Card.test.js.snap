// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a Card component, When it recieves a robot object Snapshot of the test 1`] = `
<li
  className="list-element col-4"
>
  <div
    className="card"
  >
    <img
      alt="Robot"
      className="image card-img-top"
      height="200px"
      src="http://localhost:666/"
      width="100%"
    />
    <div
      className="card-body"
    >
      <h2
        className="card-title"
      >
        Juanito
      </h2>
      <h3
        className="card-subtitle mb-2 text-muted"
      >
        created on 1994
      </h3>
      <p
        className=".card-text"
      >
        This robot has a speed of 3, endurance of 5 and it's kind of ugly.
      </p>
      <div
        className="d-flex justify-content-around"
      >
        <button
          className="btn btn-primary"
        >
          Edit
        </button>
        <button
          className="btn btn-danger"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</li>
`;
